base/range.jl

```diff
@@ -871,13 +871,22 @@ end
 
 median(r::Range{<:Real}) = mean(r)
 
-function in(x, r::Range)
-    n = step(r) == 0 ? 1 : round(Integer,(x-first(r))/step(r))+1
-    n >= 1 && n <= length(r) && r[n] == x
+function _in_range(x, r::Range)
+    if step(r) == 0
+        return !isempty(r) && first(r) == x
+    else
+        n = round(Integer, (x - first(r)) / step(r)) + 1
+        return n >= 1 && n <= length(r) && r[n] == x
+    end
 end
+in(x::Real, r::Range{<:Real}) = _in_range(x, r)
+# This method needs to be defined separately since -(::T, ::T) can be implemented
+# even if -(::T, ::Real) is not
+in(x::T, r::Range{T}) where {T} = _in_range(x, r)
 
 in(x::Integer, r::AbstractUnitRange{<:Integer}) = (first(r) <= x) & (x <= last(r))
-in(x, r::Range{T}) where {T<:Integer} =
```

Comment:
Well, let's say a runtime trait: 
```
in(x, r::Range{T}) where {T<:Integer} = isinteger(x) ?
    _fast_real_method_below(x, r) :
    (x isa Real ? false : _slow_generic_iterator_method(x, r))
```
where `_fast_real_method_below` refers to `in(x::Real, r::Range{T}) where {T<:Integer}` defined below, where the `isinteger` test can be removed.
This would not work well for e.g. `1+0im`, which is `isinteger` but not `Real`; then in `_fast_real_method_below`, the use of `x` can be replaced by `Integer(x)`.